primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,635 @@
+		for (key, val) in self.changes.iter_mut().filter(|(k, v)| predicate(k, v)) {
+			val.set(None, insert_dirty(&mut self.dirty_keys, key.to_owned()), at_extrinsic);
+		}
+	}
+
+	/// Get a list of all changes as seen by current transaction.
+	pub fn changes(&self) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
+		self.changes.iter()
+	}
+
+	/// Get the change that is next to the supplied key.
+	pub fn next_change(&self, key: &[u8]) -> Option<(&[u8], &OverlayedValue)> {
+		use std::ops::Bound;
+		let range = (Bound::Excluded(key), Bound::Unbounded);
+		self.changes.range::<[u8], _>(range).next().map(|(k, v)| (&k[..], v))
+	}
+
+	/// Consume this changeset and return all committed changes.
+	///
+	/// Panics:
+	/// Panics if there are open transactions: `transaction_depth() > 0`
+	pub fn drain_commited(self) -> impl Iterator<Item=(StorageKey, Option<StorageValue>)> {
+		assert!(self.transaction_depth() == 0, "Drain is not allowed with open transactions.");
+		self.changes.into_iter().map(|(k, mut v)| (k, v.pop_transaction().value))
+	}
+
+	/// Returns the current nesting depth of the transaction stack.
+	///
+	/// A value of zero means that no transaction is open and changes are comitted on write.
+	pub fn transaction_depth(&self) -> usize {
+		self.dirty_keys.len()
+	}
+
+	/// Start a new nested transaction.
+	///
+	/// This allows to either commit or roll back all changes that where made while this
+	/// transaction was open. Any transaction must be closed by one of the aforementioned
+	/// functions before this overlay can be converted into storage changes.
+	///
+	/// Changes made without any open transaction are comitted immediatly.
+	pub fn start_transaction(&mut self) {
+		self.dirty_keys.push(Default::default());
+	}
+
+	/// Rollback the last transaction started by `start_transaction`.
+	///
+	/// Any changes made during that transaction are discarded.
+	///
+	/// Panics:
+	/// Will panic if there is no open transaction.
+	pub fn rollback_transaction(&mut self) {
+		self.close_transaction(true);
+	}
+
+	/// Commit the last transaction started by `start_transaction`.
+	///
+	/// Any changes made during that transaction are committed.
+	///
+	/// Panics:
+	/// Will panic if there is no open transaction
+	pub fn commit_transaction(&mut self) {
+		self.close_transaction(false);
+	}
+
+	fn close_transaction(&mut self, rollback: bool) {
+		for key in self.dirty_keys.pop().expect(PROOF_DIRTY_KEYS) {
+			let value = self.changes.get_mut(&key).expect(PROOF_DIRTY_OVERLAY_VALUE);
+
+			if rollback {
+				value.pop_transaction();
+
+				// We need to remove the key as an `OverlayValue` with no transactions
+				// violates its invariant of always having at least one transaction.
+				if value.transactions.is_empty() {
+					self.changes.remove(&key);
+				}
+			} else {
+				let has_predecessor = ! if let Some(dirty_keys) = self.dirty_keys.last_mut() {
+					// Not the last tx: Did the previous tx write to this key?
+					dirty_keys.insert(key)
```

Comment:
> This breaks complexity description (the number of change to dirty keys depends on the number of transactions stacked).

The complexity is closing **one** transaction is `O(dirty_keys)` for this transaction. That holds true for every transaction.  I am reluctant to incorporate any more optimizations at this point as long as it is "good enough". Otherwise this can go on forever.

Reply:
Saying `O(dirty_keys)` is a bit misleading, when I read it first I was believing it would be related to the number of dirty keys written in the transaction but not including keys written in the child transaction.
So I was wrongly thinking that it the number of transaction would not impact the cost of any key change. Every transaction is therefore more costly than the previous one.
So I probably just misread your description.

Reply:
I think that is still in the requested performance profile. And I do not think it is misleading when I say that rollback transaction is `O(dirty_keys)`. Notice the singular. 

Also I do not think we can do any better: Keeping the transactions open is just making a time vs. space trade off.

Reply:
If by 'keeping transaction open' you are talking about my old pr, it was always clearly explained that it only get its interesting complexity by lazily cleaning transactional data and is thus more space expensive (except in corner case), that was always a known drawback it only really depends on your use case (I find this trade off good: usually we are more restrained by time than memory).
I do not either have a solution that does better without costing more memory, as I say I was just misreading the complexity (firstly understood 'dirty_keys' as dirty keys written while the transaction was the deepest one).

Reply:
I agree that this is an interesting trade off to make. However, I would rather first merge the simplest solution with an sane complexity class before resorting to more optimizations. That can quickly end in massive yak shaving otherwise. I have no problem that someones (or me) iterates on this in follow up PRs with one optimization at a time so that it is understandable and measurable.

I am grateful for your feedback. Don't get this wrong. I am just trying to get this over the finish line in a timely manner. For that reason I need to shoot down some optimization suggestions or I will be changing the implementation forever.

Your other comments regarding the client protection against misbehaving runtime will be implemented.